src/CommunityToolkit.Mvvm.SourceGenerators/ComponentModel/ObservablePropertyGenerator.Execute.cs

```diff
@@ -668,6 +670,36 @@ private static bool IsOldPropertyValueDirectlyReferenced(IFieldSymbol fieldSymbo
             return false;
         }
 
+        /// <summary>
+        /// Checks whether <see cref="MemberNotNullAttribute"/> should be used on the setter.
+        /// </summary>
+        /// <param name="fieldSymbol">The input <see cref="IFieldSymbol"/> instance to process.</param>
+        /// <param name="semanticModel">The <see cref="SemanticModel"/> instance for the current run.</param>
+        /// <returns>Whether <see cref="MemberNotNullAttribute"/> should be used on the setter.</returns>
+        private static bool GetIncludeMemberNotNullOnSetAccessor(IFieldSymbol fieldSymbol, SemanticModel semanticModel)
+        {
+            // This is used to avoid nullability warnings when setting the property from a constructor, in case the field
+            // was marked as not nullable. Nullability annotations are assumed to always be enabled to make the logic simpler.
+            // Consider this example:
+            //
+            // partial class MyViewModel : ObservableObject
+            // {
+            //    public MyViewModel()
+            //    {
+            //        Name = "Bob";
+            //    }
+            //
+            //    [ObservableProperty]
+            //    private string name;
+            // }
+            //
+            // The [MemberNotNull] attribute is needed on the setter for the generated Name property so that when Name
+            // is set, the compiler can determine that the name backing field is also being set (to a non null value).
+            return
+                fieldSymbol.Type is { NullableAnnotation: not NullableAnnotation.Annotated, IsReferenceType: true } &&
```

Comment:
I don't know whether generic types are supported in this context. If yes, consider adding a test for a generic type that is constrained to `class`. I'm not sure if this will require additional handling or if `IsReferenceType` will be sufficient.

Reply:
Oooh good catch! I'm seeing the warning even without the constraint though:

<img width="398" alt="image" src="https://user-images.githubusercontent.com/10199417/225012790-dd7bcc30-52ca-4561-ac15-fd96c39a41fb.png">

I'm thinking it might make sense to emit the attribute in all cases except if there's a `struct` constraint? 🤔
Including `unmanaged` as well, as a subset of that anyway. What do you think?

Reply:
@Sergio0694 maybe exclude `notnull` constraint as well?

Reply:
Why? If `T : notnull` it means the type is not nullable, so if you have a `T foo` field you'd definitely want the attribute on it, otherwise you'll also get a warning due to the field potentially being null, no? 🤔

I'm thinking essentially we should forward the attribute whenever the type is not annotated (which also excludes nullable value types) and not explicitly a value type (which covers reference types and unconstrained type parameters without `struct` or `unmanaged`).

Reply:
Oh, indeed. You're right.

Reply:
I've updated the generators and added a whole bunch of tests in https://github.com/CommunityToolkit/dotnet/pull/646/commits/de885cb48256c7317990183d329c19333d5cfff1, let me know what you think! 😄